resource_dispatcher/src/server.py

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+#
+
+"""Run HTTPServer for Resource Dispatcher."""
+
+import glob
+import json
+from http.server import BaseHTTPRequestHandler, HTTPServer
+from pathlib import Path
+
+import yaml
+from jinja2 import Template
+from jinja2.exceptions import TemplateSyntaxError
+
+from .log import setup_custom_logger
+
+logger = setup_custom_logger("server")
+
+
+def run_server(port: int, label: str, folder: str) -> None:
+    """Run a server for injecting resources into namespaces by given label."""
+    logger.info("Resource dispatcher service alive")
+    server = server_factory(port, label, folder)
+    logger.info(
+        f"Serving sync server forever on port: {port}, for label: {label}, on folder: {folder}"
+    )
+    server.serve_forever()
+
+
+def server_factory(controller_port: int, label: str, folder: str, url: str = "") -> HTTPServer:
+    """Return an HTTPServer populated with Handler with customised settings."""
+
+    class Controller(BaseHTTPRequestHandler):
+        def sync(self, parent, children):
+            """Return manifests which needs to be created for given state."""
+            logger.info(f"Got new request with parent: {parent} and children {children}")
+            context = {"namespace": parent.get("metadata", {}).get("name")}
+            pipeline_enabled = parent.get("metadata", {}).get("labels", {}).get(label)
+
+            if pipeline_enabled != "true":
+                logger.info(
+                    f"Namespace not in scope, no action taken (metadata.labels.{label} = {pipeline_enabled}, must be 'true')"  # noqa: E501
+                )
+                return {"status": {}, "children": []}
+
+            desired_secret_count = 1
+            desired_resources = []
+            desired_status = {
+                "resources-ready": str(len(children["Secret.v1"]) == desired_secret_count)
+            }
```

Comment:
This one is interesting.  I misunderstood this field until you used it here.  I thought it was a boolean "are my resources right?" sent to metacontroller, but it actually [sets the object's status field](https://metacontroller.github.io/metacontroller/api/compositecontroller.html#sync-hook-response) like you're doing here.  I like this addition.

When you get to charming this, keep this field in mind.  This is a lagging indicator (eg: this execution of the script returns {`status before I started working`, `resources I want to see`}.  So if we see things need updating, we're going to report that `status=bad`.  If metacontroller then calls us back a few seconds later once it built the resources we wanted, we'll respond `status=good` and all is well.  But the `CompositeController`'s [resync period](https://github.com/canonical/resource-dispatcher/pull/1/files#diff-900c2a543e4efb22c9953cba878be42fd0b65fb20f0c3cabcbc4568f6dc8e754R7) is what controls this.  10s like you have in the other PR right now is probably great (we can only be behind in status by 10s).  3600s like upstream uses would be bad.

Reply:
What could help here is the optional third data item you can return, `resyncAfterSeconds`.  What you could do here is if `resources-ready==False`, add a `resyncAfterSeconds=10` or something similar to the response.  This'll tell metacontroller that you need a resync soon, and then our statuses will not be stale even if the CompositeController's default is a long sync process

Reply:
We discussed offline.  This change was important because the script here has no control over the resync pattern that someone submits it with, so a user could (incorrecltly) deploy it with a long resync time.  

But this is now fixed by https://github.com/canonical/resource-dispatcher-image/pull/1/commits/eef10f913e701c93c26c9a3cbc2caf22488ad378 and https://github.com/canonical/resource-dispatcher-image/pull/1/commits/f23798d243b9d057789b3c2237dbe1fca3d4ca62